### ISSUE TITLE
Idea: New plugin "mpts-optimizer"

### ISSUE THREAD

**Issue by lars18th** [tsduck/tsduck#187]
Hi,

I have a new idea/suggestion to improve the muxing capabilities of the TSDuck.

I understand that TSDuck isn’t a professional muxer. However, at time, it’s the most complete software tool in the opensource area to create MPTS streams. And for this reason I feel that any enhancement will be a significant improvement for us. So, only for this reason I write here my ideas about it. I hope you don’t think they’re stupid and/or crazy.
 
Regarding the current implementation of the `merge` plugin, it does the work as expected. But two issues are still present regarding the “muxing” output:

- PCR restamping: Only PCRs from the merged stream can be restamped. So, the resulting MPTS can have incorrect PCR marks in some cases.
- PID distribution: The current algorithm writes adjacent consecutive packets. The result is then a MPTS without sufficient dispersion of the different services. To understand what I mean, look at these two examples:

This first image is the “Grid View” from DVB Inspector of a regular DVB MPTS broadcasted over a terrestrial network:

![mpts-dvb](https://user-images.githubusercontent.com/9989964/50561113-97885980-0cff-11e9-8ab3-91aa2b8fe9b8.png)

And this is the same view from a similar MPTS created using TSDuck:

![mpts-tsduck](https://user-images.githubusercontent.com/9989964/50561116-a111c180-0cff-11e9-8bd8-7e0c59a9ddec.png)

I’m sure you can see the difference!

So, here my suggestion about a new `mpts-optimizer` plugin:

- This plugin will process DVB MPTS streams with a CBR.
- The first task can be to restamp all PCR based on the different services present in the MPTS.
- Another task could be to dispersate (aka obfuscate) packets from different pids. One simple strategy can be: for packets of the same pid between two PCR marks, not move any packet with a PCR mark, but for the reset try to relocate to another position. With the restriction of maximize the distance between consecutive packets. And this will be done for all pids.
- Furthermore, a final task can be to check the conformance of the MPTS, and especially for the bitrate of the resulting stream.

What you think about this idea?

**Comment by lelegard** [comment]
Good idea. But this will require a new dedicated tool since `tsp` plugins process packets one by one. A plugin does not have a global view of the packet buffer and consequently cannot reorder packets. The idea of a MUX tool has been in the air for many years but I never found enough time to develop it.

**Comment by lars18th** [comment]
I'm glad you find the idea interesting.
But I'm disappointed the TSDuck can't handle it.

In any case, thank you for this impresive tool! :+1:

**Comment by lelegard** [comment]
The `merge` plugin could improve the distribution of merged packets when there are much more stuffing in the main stream than packets to be merged. I have some ideas. I will continue to think about it and implement it later.

**Comment by lars18th** [comment]
> The `merge` plugin could improve the distribution of merged packets when there are much more stuffing in the main stream than packets to be merged. I have some ideas. I will continue to think about it and implement it later.

Great!

However, I'm not sure if it will can do a very good dispersion. But _any_ improvement in this way will be great.

And regarding the `merge` plugin and the PCR restamping: Can you consider a "remerge" functionality? The idea is more or less this: from a source MPTS, "merge" it with a full empty NULL TS and recalculate the PCR of _all_ services without touching any other stuff. In this case no async (subprocess) mode is required, as only the source stream is required. What you think?

Regards.

**Comment by samerajarmeh** [comment]
thanks for your idea 
i use mptsd for more than a year with pcr restamp but sometime audio lost from streams 
its open and free software but need more to be stable software
https://github.com/gfto/mptsd
if tsduck support MPTS will be complete MPEG software
i hope it will support it..
thanks for all ..

**Comment by lars18th** [comment]
Hi @samerajarmeh ,

I do a lot of work with the `mptsd` tool (see my still _pending_ PRs: https://github.com/gfto/mptsd/pulls).
However, I don't like it, as it generates a lot of errors.

For this reason I suggest to improve the TSDuck in this way. I feel that with _some_ pid dispersion, and _global_ PCR restamping, then we can get a _sufficient_ robust muxer.

Regards.

**Comment by lars18th** [comment]
> A plugin does not have a global view of the packet buffer and consequently cannot reorder packets. 

Hi @lelegard ,

Perhaps some FIFO strategy can solve the problem for a very particular case (not as regular behaviour, as it has severy overhead penalties):

- To reorder packets, the plugin can use an internal FIFO. At start it copies new incoming packets to the FIFO, and for each packet it frees the packet to a NULL. Then when the FIFO is full, it can start to processes in this way: for each incoming packet, copy it to the FIFO and replace it with the top in the FIFO. With this solution you can do _any_ processing in the internal FIFO in each loop.
- The drawbacks are relevant: non copy-on-write, more memory required, and added delay because the internal buffer.

I'm not sure if you accept to include such type of plugins, as I feel you prefer to maintain TSDuck completely away from multiple buffer copies.

These are just my five cents!

**Comment by SonnyWalkman** [comment]
Hello,
I agree with **lars18th**, I've found mptsd  bit flaky produces random errors. not sure why and how?
I'll support a professional muxer.
I use gstreamer framework a far bit which allows queues to decouple elements in a pipeline. Same idea as a FIFO using shared memory.  Check out link below:

https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gstreamer-plugins/html/gstreamer-plugins-queue.html

Maybe worth looking in on the source on how it does it's magic (written in 'C')

With the cost of CPU power and RAM of these days. I few milliseconds additional delay in FIFO processing can been seen as a  small price to pay over reliability in my view.

I'm new to TSduck on Linux and find it one of the best tools out there .. 

Rob
